plugins/out_ws_nats/ws_nats.c

```diff
@@ -0,0 +1,434 @@
+
+        for (i = 0; i < map_size; i++) {
+            m_key = map.via.map.ptr[i].key;
+            m_val = map.via.map.ptr[i].val;
+
+            msgpack_pack_object(&mp_pck, m_key);
+            msgpack_pack_object(&mp_pck, m_val);
+        }
+    }
+    msgpack_unpacked_destroy(&result);
+
+    out_buf = flb_msgpack_raw_to_json_sds(mp_sbuf.data, mp_sbuf.size);
+    msgpack_sbuffer_destroy(&mp_sbuf);
+
+    if (!out_buf) {
+        return -1;
+    }
+
+    *out_json = out_buf;
+    *out_size = flb_sds_len(out_buf);
+
+    return 0;
+}
+
+static int cb_ws_init(struct flb_output_instance *ins,
+                      struct flb_config *config, void *data)
+{
+    struct flb_out_ws_nats *ctx = NULL;
+
+    ctx = flb_ws_nats_conf_create(ins, config);
+    if (!ctx) {
+        return -1;
+    }
+
+    ctx->handshake = 1;
+    ctx->last_input_timestamp = time(NULL);
+    flb_output_set_context(ins, ctx);
+    return 0;
+}
+
+static int cb_ws_exit(void *data, struct flb_config *config)
+{
+    struct flb_out_ws_nats *ctx = data;
+    flb_ws_conf_destroy(ctx);
+    return 0;
+}
+
+static void cb_ws_flush(const void *data, size_t bytes,
+                        const char *tag, int tag_len,
+                        struct flb_input_instance *i_ins,
+                        void *out_context,
+                        struct flb_config *config)
+{
+    int ret = -1;
+    size_t bytes_sent;
+    char *connect_proto;
+    size_t connect_proto_len;
+    size_t json_len;
+    flb_sds_t json_msg;
+    char *request;
+    int req_len;
+    flb_sds_t json = NULL;
+    struct flb_upstream *u;
+    struct flb_upstream_conn *u_conn;
+    struct flb_out_ws_nats *ctx = out_context;
+    time_t now;
+
+    /* Get upstream context and connection */
+    u = ctx->u;
+    u_conn = flb_upstream_conn_get(u);
+
+    if (!u_conn) {
+        flb_error("[out_ws_nats] no upstream connections available to %s:%i", u->tcp_host, u->tcp_port);
+        ctx->handshake = 1;
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+    now = time(NULL);
+
+    flb_debug("[out_ws_nats] interval is  %ld and handshake is %d", now - ctx->last_input_timestamp, ctx->handshake);
+    if ((now - ctx->last_input_timestamp > ctx->idle_interval) && (ctx->handshake == 0)) {
+        ctx->handshake = 1;
+        flb_upstream_conn_release(u_conn);
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+    ctx->last_input_timestamp = now;
+
+    if (ctx->handshake == 1) {
+        /* Handshake with websocket server*/
+        flb_info("[out_ws_nats] handshake for ws");
+        ret = flb_ws_handshake(u_conn, ctx);
+        if (ret == -1) {
+            flb_upstream_conn_release(u_conn);
+            FLB_OUTPUT_RETURN(FLB_RETRY);
+        }
+        ctx->handshake = 0;
+
+        /* Send CONNECT whenever the handshake has to happen  */
+        connect_proto = flb_malloc(sizeof(NATS_CONNECT));
+        connect_proto_len = snprintf(connect_proto, sizeof(NATS_CONNECT), NATS_CONNECT);
```

Comment:
I think these values should be allocated only once since NATS_CONNECT is a static string.

Is `connect_proto` released after using?